Python/Product/VSInterpreters/Interpreter/WorkspaceInterpreterFactoryProvider.cs

```diff
@@ -215,21 +217,35 @@ class WorkspaceInterpreterFactoryProvider : IPythonInterpreterFactoryProvider, I
         }
 
         private static IEnumerable<PythonInterpreterInformation> FindInterpretersInSubFolders(string workspaceFolder) {
-            foreach (var dir in PathUtils.EnumerateDirectories(workspaceFolder, recurse: false)) {
-                var file = PathUtils.FindFile(dir, "python.exe", depthLimit: 1);
-                if (!string.IsNullOrEmpty(file)) {
-                    yield return CreateEnvironmentInfo(file);
-                }
+            foreach (var directory in PathUtils.EnumerateDirectories(workspaceFolder, recurse: false)) {
+                yield return FindInterpreterInFolder(directory);
             }
         }
 
-        private void OnFileCreated(object sender, FileSystemEventArgs e) {
+        private static PythonInterpreterInformation FindInterpreterInFolder(string folder) {
+            var file = PathUtils.FindFile(folder, "python.exe", 1);
+            if (!string.IsNullOrEmpty(file)) {
+                return CreateEnvironmentInfo(file);
+            }
+
+            return null;
+        }
+
+        private void OnFileCreatedDeletedRenamed(object sender, FileSystemEventArgs e) {
             lock (_factories) {
                 try {
-                    if (string.Compare(Path.GetFileName(e.FullPath), "python.exe", StringComparison.OrdinalIgnoreCase) == 0) {
-                        _pythonExecutableCreated = true;
+                    if (Directory.Exists(e.FullPath) && e.ChangeType == WatcherChangeTypes.Renamed) {
+                        _refreshPythonInterpreters = true;
```

Comment:
We need the additional logic here that checks if the folder being renamed affects any of the current interpreter factories.

Without it, we're going to refresh all factories when any workspace folder is renamed.

Reply:
Updated